### ISSUE TITLE
API for filtering / dropping?

### ISSUE THREAD

**Issue by erik-stephens** [sematext/logagent-js#22]
I can't find anything in the docs or code to show how to drop a message.  Is that supported?  If not and something you'd like to support, any thoughts on how you'd like to support that functionality?  Some options that come to mind:

- Allow transform functions to declare the message as filtered (eg `parsed.filter(bool)`)?
- Support something similar to `transform` (eg `filtered`) that accepts a function that should return true if message should be dropped?

**Comment by megastef** [comment]
+1

The simple version is to specify an 'inputFilter' for a pattern as regular expression. When the regex matches the parsing will be executed, else skipped. 

```
patterns:
 - 
  sourceName: !!js/regexp /testFilter/i
  match:
    - type: testDropFilter
      inputFilter: !!js/regexp /warning|error/i
      regex: !!js/regexp (.*)
      fields: [message:string]
```

Regarding the filter() function, my suggestion is a pattern specific function, which gets the parsed object and the pattern object to define custom properties (e.g. regex helpers) - called after all transform functions, in case one likes to filter on some parsed fields. 

Example, which simulates the behaviour of the above "inputFilter": 
```
patterns:
 - 
  sourceName: !!js/regexp /testFilterFunction/i
  match:
    - type: testDropFilter
      regex: !!js/regexp (.*)
      fields: [message:string]
      myCustomFilterProperty: !!js/regexp /warning|error/i
      filter: !!js/function >
        function (parsed, pattern) {
           // simplest filter function as example :) just return regex test ....
           return pattern.myCustomFilterProperty.test(parsed.message)
        }
``` 

Declare a message to be dropped is another good idea, but we should use variable name, which not common as "filter" - e.g. ```logagentDropMessage=true````, which could be used in  globalTransform() or transform() function. 

I'm going to submit this change for logagent 1.x branch (which can be released to NPM until the work for 2.0 is finished).

**Comment by otisg** [comment]
Out of curiosity, @erik-stephens , what use-case for dropping events do you have in mind?
@megastef note about terminology in Logagent in general: it it log, event, or message?

**Comment by erik-stephens** [comment]
The most common filter for us is to only ship higher severity events.  Sometimes that threshold is based on parsed fields such as the process emitting the event (drop all debug unless process=foo).  We have other less general use cases, which is why I like the ability to define that via a user-defined function and not something like a regex.  Compared to  logstash & filebeat, logagent is more attractive since I can write simple functions to process events (no fighting a DSL).

**Comment by megastef** [comment]
@otis I would stay with the "message" term. 
@erik-stephens thanks for the feedback! 

Published on npm v1.2.23, patterns  can use following options: 
- inputFilter - regex, only matched RAW messages (could be multiline log message) are parsed
- inputDrop - regex, drops matched RAW messages (don't parse)
- filter - JS function to return true to pass the parsed object, false to drop the parsed object

The advantage of inputFilter/inputDrop is that it could save CPU cycles for parsing, while the filter function could use any field from the parsed message. 
```
patterns:
 - 
  sourceName: !!js/regexp /testFilterFunction/i
  match:
    - type: testDropFilter
      # parse only messages with "stats" in the RAW log message 
      inputFilter: !!js/regexp /stats/i
      # drop messages with "debug" in the RAW log message 
      inputDrop: !!js/regexp /debug/i
      regex: !!js/regexp (.*)
      fields: [message:string]
      # define a custom property, which can be used in filter function
      myCustomFilterProperty: !!js/regexp /.*/i
      # create a custom filter function, to filter after parsing 
      filter: !!js/function >
        function filter (parsed, pattern) {
          return pattern.myCustomFilterProperty.test(parsed.message)
        }
```

I leave the issue open, as this should be included in the next release (master branch) as well.

**Comment by megastef** [comment]
Added to 2.x as it is a new feature, to mention in release notes.
